### DISCUSSION TITLE
Can I update JSON field without replace other data?

### DISCUSSION THREAD

**Discussion by szulcus** [supabase/supabase#14174] [Questions] [answered]
I have this data in `progress` field:
```json
{ "test-book": { "test-unit": { "test-part": 13, "test-part-2": 22 } } }
```
and I want to have this data:
```json
{ "test-book": { "test-unit": { "test-part": 10, "test-part-2": 22 } } }
```
but when I do:
```ts
const { data, error } = await supabase.from('profiles').update({
	progress: { 'test-book': { 'test-unit': { 'test-part': 10 } } },
}).eq('id', 'ec3ca0df-bb99-465d-93ec-054153b19d7a').select();
console.log({ data, error });
```
I will have:
```json
{ "test-book": { "test-unit": { "test-part": 10 }
```

Or maybe there is a simpler and faster way to update such nested fields in JSON, as in the case of retrieving them (https://supabase.com/docs/guides/database/json)?

**Comment by GaryAustin1** [comment, marked as answer]
Not with the API.  https://supabase.com/docs/reference/javascript/update
Postgres offers some [operators](https://supabase.com/docs/guides/database/json#query-the-jsonb-data) for working with JSON data. Currently, it is only possible to update the entire JSON document.

An rpc call to a Postgres function is only way.

**Reply by szulcus** [reply]
Thank you for your answer. Is work already underway? Maybe it's worth considering such a feature? If this is to be a true alternative to Firebase, it seems to me necessary.

**Reply by GaryAustin1** [reply]
The API Supabase uses for database operations is PostgREST.  This issue there has tracked this for a long time.  Don't see much progress.
https://github.com/PostgREST/postgrest/issues/465 

Because Supabase lets you execute SQL directly you can do alot of things with Postgres functions using SQL and all the json operators it offers.
You would pass in your info to be updated and then in the function replace or add using JSON operators.

A supabase discussion that seems close: https://github.com/orgs/supabase/discussions/1570
A general discussion: https://aaronbos.dev/posts/update-json-postgresql

There are lots of examples of using json operators in Postgres SQL outthere.

**Comment by matiasbattocchia** [comment]
Just in case it helps someone. This function will deep merge the target jsonb in `column_name` with `object`.

```js
    const { error } = await client.rpc("merge_update_jsonb", {
      table_name: "table",
      column_name: "column",
      row_id: "id",
      object: { ... },
    });
```

```sql
create or replace function public.merge_update_jsonb(target jsonb, path text[], object jsonb) returns jsonb as $$
declare
  i int;
  key text;
  value jsonb;
begin
  case jsonb_typeof(object) -- object, array, string, number, boolean, and null
    when 'object' then
      
      if jsonb_typeof(target #> path) <> 'object' or target #> path is null then
        target := jsonb_set(target, path, '{}', true);
      end if;

      for key, value in select * from jsonb_each(object) loop
        target := merge_update_jsonb(target, array_append(path, key), value); 
      end loop;
    when 'array' then
      if jsonb_typeof(target #> path) <> 'array' or target #> path is null then
        target := jsonb_set(target, path, '[]', true);
      end if;

      i := 0;

      for value in select * from jsonb_array_elements(object) loop
        target := merge_update_jsonb(target, array_append(path, i::text), value);
        i := i + 1;
      end loop;
    else
      target := jsonb_set(target, path, object, true);
  end case;
  
  return target;
end;
$$ language plpgsql immutable;

create or replace function public.merge_update_jsonb(table_name text, column_name text, row_id uuid, object jsonb) returns jsonb as $$
declare
  target jsonb;
begin
  execute format('select %I from %I WHERE id = %L', column_name, table_name, row_id) into target;
  
  target := merge_update_jsonb(target, '{}', object);

  execute format('update %I set %I = %L::jsonb where id = %L', table_name, column_name, target::text, row_id);

  return target;
end;
$$ language plpgsql;
```

It seems to work ok but it has not been battle tested yet. Use it at your own risk.

**Comment by softmarshmallow** [comment]
Will this be ever available on postgrest without rpc?

**Comment by mvp-studio-repo** [comment]
raising this case again, still on the go it's seems
